developer-docs/architecture/type-system.md

```diff
@@ -0,0 +1,260 @@
+types. Consider a variant of the `Array<string>` example from [`inputShape(T)`](#inputshapet):
+in order to produce an array where each element may be an output, we need to use the type
+`Output<Array<Output<string>>>`.
+
+The `outputShape` type constructor defines an algorithm for producing these sorts of values.
+
+```rust
+fn outputShape(T) {
+	match T {
+		_ => Output<T>,
+		Array<U> => Output<Array<outputShape(U)>>,
+		Map<U> => Output<Map<outputShape(U)>>,
+		Union<...U> => Union<map(...U, u => outputShape(u))>,
+		Promise<U> => Output<U>,
+		Output<U> => Output<U>,
+		Object<...P> => Output<Object<map(...P, (name, u) => (name, outputShape(u)))>>
+	}
+}
+```
+
+Resource output properties often use output-shaped types.
+
+Projecting output-shaped is a bit unwieldy, as values of these types often require a great
+deal of unwrapping as nesting depth increases.
+
+Instead of projecting these types in their fully-elaborated form, the various language SDKs
+tend to opt to project them as a simple `Output<T>` while using an internal representation
+for the concrete value that includes distinguished unknown values. This approach lets the
+SDKs to allow e.g. lifted property and element access into partially-known composite
+values. For example, the Node SDK will allow the user to access an element of an
+`Output<[]string>` via a proxied index operator even if some elements of the array are
+unknown, though it will not allow the user to access the entire value via `apply`.
+
+## `Output<T>` Combinators
+
+The rules described for working with `Output<T>` metadata--[dependencies], [unknowns], and
+[secrets]--require special bookkeeping on the part of the consumer. There are three
+primitive combinators that aid in this bookkeeping: [`apply`](#applyt-u), [`all`](#allt),
+and [`unwrap`](#unwrapt).
+
+### `apply<T, U>`
+
+The `apply` API allows its caller to access the concrete value of an `Output<T>` within
+the context of a caller-supplied callback.
+
+`apply` trivially obeys the `Output<T>` rules for [dependencies], [unknowns], and
+[secrets]:
+
+- the dependencies of the `Output<T>` argument are propagated to the result
+- if the `Output<T>` argument is unknown, the callback is not run and the result is unknown
+- if the `Output<T>` argument is secret, the result is secret
+
+Note that the argument for `U` may itself be an `Output<V>`, in which case the return type
+of `apply` will be `Output<Output<V>>`. The result can be unwrapped using the
+[`unwrap`](#unwrapt) combinator. A language SDK may opt to automatically unwrap such values
+if its type system is flexible enough to express the unwrapping.
+
+This API is morally equivalent to Javascript's `Promise.then` API, but with `Output<>`s in
+the place of `Promise<>`s:
+
+```typescript
+class Output<T> {
+	public apply<U>(func: (t: T) => U): Output<U> {
+		...
+	}
+}
+```
+
+### `all<T...>`
+
+The `all` API combines multiple heterogenous outputs into a single array output. The
```

Comment:
Not always an array, sometimes a mapping (at least in python, and I thought also in TS but unsure of this)

Reply:
Yeah--the language SDKs generally implement a more flexible version of `all`.

To be frank, I'm not entirely convinced that the type system docs even need `all`, as I _think_ it can be described in terms of `apply` and `unwrap`. It seemed like a nice thing to have, though, and given that it's here it should probably include the `Map<>` case (or the `Map<>` case should be described in a separate combinator).

Reply:
It does seem nice to have, agreed.